## Data Collection

In [1]:
import csv
from bs4 import BeautifulSoup
import requests
import pandas as pd
import time
time.sleep(2)
import warnings
warnings.filterwarnings('ignore')

In [2]:
urls = []
url1 = 'https://www.imdb.com/title/tt0038650/reviews/?ref_=tt_ql_urv'
url2 = 'https://www.imdb.com/title/tt0078788/reviews/?ref_=tt_ql_urv' 
url3 = 'https://www.imdb.com/title/tt0096870/reviews/?ref_=tt_ql_urv' 
url4 = 'https://www.imdb.com/title/tt0110978/reviews/?ref_=tt_ql_urv' 
url5 = 'https://www.imdb.com/title/tt10366460/reviews/?ref_=tt_ql_urv'
url6 = 'https://www.imdb.com/title/tt2802144/reviews/?ref_=tt_ql_urv'

In [3]:
urls.append(url1) 
urls.append(url2) 
urls.append(url3) 
urls.append(url4) 
urls.append(url5)
urls.append(url6)

In [4]:
content = []

for url in urls:
    page = requests.get(url, timeout=2.50) 
    page_content = page.content
    soup = BeautifulSoup(page_content, 'html.parser')
    content.append(soup.find_all('div', class_= 'review-container'))


In [5]:
print(content)

[[<div class="review-container">
<div class="lister-item-content">
<div class="ipl-ratings-bar">
<span class="rating-other-user-rating">
<svg class="ipl-icon ipl-star-icon" fill="#000000" height="24" viewbox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
<path d="M0 0h24v24H0z" fill="none"></path>
<path d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"></path>
<path d="M0 0h24v24H0z" fill="none"></path>
</svg>
<span>8</span><span class="point-scale">/10</span>
</span>
</div>
<a class="title" href="/review/rw1089030/"> Very good and very overrated
</a> <div class="display-name-date">
<span class="display-name-link"><a href="/user/ur2467618/">MartinHafer</a></span><span class="review-date">26 May 2005</span>
</div>
<div class="content">
<div class="text show-more__control">This is a very good movie. Period. A "classic", I don't think so. This is a movie whose actual quality has been exceeded by its hype--though NOT because the produc

In [6]:
movie = pd.DataFrame(columns=['Review','Rating'])

In [7]:
review = [] 
rating = [] 
count = 0
for cc in content:
    for c in cc:
        count+= 1

        print('\nMovie review ', count)
        #Get review.
        str = c.find_all('a', attrs={'class':'title'}) 
        rReview =''
        for s in str:
            #print('Review is: ',s.get_text())
            rReview = s.get_text()

        #Get rating.
        ratings = c.find_all('span', attrs={'class':''}) 
        rVal = []
        for r in ratings:
            str1 = r.get_text().strip() 
            rVal.append(str1)

        val = rVal[0]

        if(len(val) > 2):
            continue 
        else:
            review.append(rReview) 
            rating.append(val)

            print('Review: ', rReview) 
            print('Rating: ',val)

movie['Review'] = review 
movie['Rating'] = rating



Movie review  1
Review:   Very good and very overrated

Rating:  8

Movie review  2
Review:   One of those quintessential Christmas movies....

Rating:  10

Movie review  3
Review:   "To My Big Brother George, The Richest Man In Town"

Rating:  10

Movie review  4
Review:   Great Christmas (and all year round) movie

Rating:  9

Movie review  5

Movie review  6
Review:   Unabashed Sentimental Capra

Rating:  9

Movie review  7
Review:   A Wonderful Film with a Cutting Edge

Rating:  10

Movie review  8

Movie review  9
Review:   Each man's life touches so many others.

Rating:  10

Movie review  10
Review:   Small town life

Rating:  7

Movie review  11
Review:   One of the Most Beautiful Movies of the Cinema History Ever

Rating:  10

Movie review  12

Movie review  13
Review:   Life - is what you make it (for you and for ...)

Rating:  8

Movie review  14
Review:   "Each man's life touches so many other lives".

Rating:  10

Movie review  15
Review:   Did anyone read Maureen Dowd's 

In [8]:
movie.head()

Review Rating
0                     Very good and very overrated\n      8
1   One of those quintessential Christmas movies....     10
2   "To My Big Brother George, The Richest Man In...     10
3       Great Christmas (and all year round) movie\n      9
4                      Unabashed Sentimental Capra\n      9

In [9]:
movie.shape

(130, 2)

In [10]:
movie.to_csv('Sharanpreet singh.csv', index=False)

## Text Processing and Analysis

In [11]:
import string
import re
#import nltk
#nltk.download()
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.model_selection import train_test_split


In [12]:
textFeatures = movie['Review'].copy() 
textFeatures.shape

(130,)

In [13]:
#Preparing text for Wordcloud
text = []
for t in textFeatures: 
    text.append(t)

all_text = ', '.join(t for t in text)
#print(all_text)
print(len(all_text))


5400


In [15]:
from os import path
from PIL import Image
# from wordcloud import WordCloud, STOPWORDS, ImageColorGenerator

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline


In [16]:
from nltk.corpus import stopwords

In [17]:
# Create stopword list
stopwords = set(stopwords)
stopwords.update(["br", "im", "thats"]) #"im","lol","Xa","film"])

# Generate a word cloud image
wordcloud = WordCloud(stopwords=stopwords, background_color="white").generate(all_text)

# Display the image
plt.imshow(wordcloud, interpolation='bilinear') plt.axis("off")
plt.show()
#save the generated image to a fileb
#wordcloud.to_file("wordcloud_cb_all.png")


## Sentiment Identification using VADER

In [18]:
import nltk
nltk.download('vader_lexicon')

from nltk.sentiment.vader import SentimentIntensityAnalyzer


[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\akash\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [19]:
sid = SentimentIntensityAnalyzer() 
c = 0
for t in text: 
    c+=1
    print(c, t)
    ss = sid.polarity_scores(t) 
    print(ss)

    if(ss['compound'] >= 0.05): 
        print('positive')

    elif(ss['compound'] <= -0.05): 
        print('negative')

    else:
        print('neutral') 
    print('\n')


1  Very good and very overrated

{'neg': 0.0, 'neu': 0.556, 'pos': 0.444, 'compound': 0.4927}
positive


2  One of those quintessential Christmas movies....

{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}
neutral


3  "To My Big Brother George, The Richest Man In Town"

{'neg': 0.0, 'neu': 0.726, 'pos': 0.274, 'compound': 0.5267}
positive


4  Great Christmas (and all year round) movie

{'neg': 0.0, 'neu': 0.594, 'pos': 0.406, 'compound': 0.6249}
positive


5  Unabashed Sentimental Capra

{'neg': 0.0, 'neu': 0.465, 'pos': 0.535, 'compound': 0.3182}
positive


6  A Wonderful Film with a Cutting Edge

{'neg': 0.183, 'neu': 0.366, 'pos': 0.451, 'compound': 0.4939}
positive


7  Each man's life touches so many others.

{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}
neutral


8  Small town life

{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}
neutral


9  One of the Most Beautiful Movies of the Cinema History Ever

{'neg': 0.0, 'neu': 0.705, 'pos': 0.295, 'compound': 0.6

## Sentiment Classification using Machine Learning

In [20]:
label = []

for r in movie['Rating']: 
    r = int(r)
    if (r>5):
        label.append('1') #Positive
    elif(r<5):
        label.append('-1') #Negative
    elif(r==5):
        label.append('0') #Netural

movie['class-label'] = label


In [21]:
movie['class-label'].value_counts()

1     90
-1    36
0      4
Name: class-label, dtype: int64

In [22]:
movie = movie[movie['class-label']!='0']

In [23]:
movie['class-label'].value_counts()

1     90
-1    36
Name: class-label, dtype: int64

In [24]:
textFeatures = movie['Review'].copy() 
textFeatures.shape

(126,)

In [25]:
import nltk
nltk.download('punkt')

# Stemming using TextBlob library for stemming
from textblob import TextBlob


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\akash\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [26]:
def textblob_tokenizer(input_str):
    blob = TextBlob(input_str.lower()) 
    tokens = blob.words
    words = [token.stem() for token in tokens]
    return words


In [27]:
#Example text
print(textblob_tokenizer('Q: studed studing!!! I miss uuuu! It&#039;s'))

['q', 'stude', 'stude', 'i', 'miss', 'uuuu', 'it', '039', 's']


In [28]:
#Transforming text dataset in to two matrix representations.

#countvectorizer convers each review into a vector based on the word count.
countvectorizer = CountVectorizer(analyzer= 'word', stop_words= 'english',
                                    tokenizer=textblob_tokenizer)

#convers text into a vector based on tf-idf weighting scheme.
tfidfvectorizer = TfidfVectorizer(analyzer= 'word', stop_words= 'english',
                                    tokenizer=textblob_tokenizer)


In [29]:
count_matrix = countvectorizer.fit_transform(textFeatures) 
tfidf_matrix = tfidfvectorizer.fit_transform(textFeatures)

In [30]:
print(tfidf_matrix.shape) 
print(count_matrix.shape)

(126, 311)
(126, 311)


## Build ML models

In [31]:
features_train, features_test, labels_train, labels_test = train_test_split(tfidf_matrix,movie['class-label'], test_size=0.3,random_state=8) 
print(features_train.shape, features_test.shape, labels_train.shape, labels_test.shape)


(88, 311) (38, 311) (88,) (38,)


In [32]:
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.metrics import accuracy_score


In [33]:
#SVM classifier

from sklearn.svm import SVC
print("\nEvaluation for SVM \n")
svc = SVC(kernel='sigmoid', gamma=1.0) 
svc.fit(features_train, labels_train) 
prediction = svc.predict(features_test)
acc = accuracy_score(labels_test,prediction) 
print('Accuracy:',acc)

from sklearn.metrics import precision_score
prec = precision_score(labels_test,prediction, average='weighted') 
print('Precision:',prec)

from sklearn.metrics import recall_score
recall = recall_score(labels_test,prediction, average='weighted') 
print('Recall:',recall)

from sklearn.metrics import f1_score
f1 = f1_score(labels_test,prediction, average='weighted') 
print('F-1 measure: ', f1)
print('\nConfusion Matrix:\n')
print(confusion_matrix(labels_test, prediction))
print(classification_report(labels_test, prediction))
#print(prediction)



Evaluation for SVM 

Accuracy: 0.7368421052631579
Precision: 0.6842105263157895
Recall: 0.7368421052631579
F-1 measure:  0.6655701754385964

Confusion Matrix:

[[ 1  9]
 [ 1 27]]
              precision    recall  f1-score   support

          -1       0.50      0.10      0.17        10
           1       0.75      0.96      0.84        28

    accuracy                           0.74        38
   macro avg       0.62      0.53      0.51        38
weighted avg       0.68      0.74      0.67        38



In [34]:
#Decision Tree

print("\nEvaluation for Decision Tree \n")
from sklearn.tree import DecisionTreeClassifier

dtree = DecisionTreeClassifier()
dtree.fit(features_train, labels_train) 
prediction = dtree.predict(features_test)
acc = accuracy_score(labels_test,prediction) 
print('Accuracy: ',acc)
prec = precision_score(labels_test,prediction, average='weighted') 
print('Precision: ',prec)
recall = recall_score(labels_test,prediction, average='weighted') 
print('Recall: ',recall)
f1 = f1_score(labels_test,prediction, average='weighted') 
print('F-1 measure: ',f1)
print('\nConfusion Matrix:\n')
print(confusion_matrix(labels_test, prediction))
print(classification_report(labels_test, prediction))



Evaluation for Decision Tree 

Accuracy:  0.5
Precision:  0.7076923076923077
Recall:  0.5
F-1 measure:  0.5156794425087107

Confusion Matrix:

[[ 8  2]
 [17 11]]
              precision    recall  f1-score   support

          -1       0.32      0.80      0.46        10
           1       0.85      0.39      0.54        28

    accuracy                           0.50        38
   macro avg       0.58      0.60      0.50        38
weighted avg       0.71      0.50      0.52        38



### Challenge Task

In [35]:
features_train, features_test, labels_train, labels_test = train_test_split(count_matrix,movie['class-label'], test_size=0.3,random_state=8) 
print(features_train.shape, features_test.shape, labels_train.shape, labels_test.shape)


(88, 311) (38, 311) (88,) (38,)


In [36]:
#SVM classifier

from sklearn.svm import SVC
print("\nEvaluation for SVM \n")
svc = SVC(kernel='sigmoid', gamma=1.0) 
svc.fit(features_train, labels_train) 
prediction = svc.predict(features_test)
acc = accuracy_score(labels_test,prediction) 
print('Accuracy:',acc)

from sklearn.metrics import precision_score
prec = precision_score(labels_test,prediction, average='weighted') 
print('Precision:',prec)

from sklearn.metrics import recall_score
recall = recall_score(labels_test,prediction, average='weighted') 
print('Recall:',recall)

from sklearn.metrics import f1_score
f1 = f1_score(labels_test,prediction, average='weighted') 
print('F-1 measure: ', f1)
print('\nConfusion Matrix:\n')
print(confusion_matrix(labels_test, prediction))
print(classification_report(labels_test, prediction))
#print(prediction)



Evaluation for SVM 

Accuracy: 0.868421052631579
Precision: 0.8652954224642065
Recall: 0.868421052631579
F-1 measure:  0.866112650046168

Confusion Matrix:

[[ 7  3]
 [ 2 26]]
              precision    recall  f1-score   support

          -1       0.78      0.70      0.74        10
           1       0.90      0.93      0.91        28

    accuracy                           0.87        38
   macro avg       0.84      0.81      0.82        38
weighted avg       0.87      0.87      0.87        38



In [37]:
#Decision Tree

print("\nEvaluation for Decision Tree \n")
from sklearn.tree import DecisionTreeClassifier

dtree = DecisionTreeClassifier()
dtree.fit(features_train, labels_train) 
prediction = dtree.predict(features_test)
acc = accuracy_score(labels_test,prediction) 
print('Accuracy: ',acc)
prec = precision_score(labels_test,prediction, average='weighted') 
print('Precision: ',prec)
recall = recall_score(labels_test,prediction, average='weighted') 
print('Recall: ',recall)
f1 = f1_score(labels_test,prediction, average='weighted') 
print('F-1 measure: ',f1)
print('\nConfusion Matrix:\n')
print(confusion_matrix(labels_test, prediction))
print(classification_report(labels_test, prediction))



Evaluation for Decision Tree 

Accuracy:  0.47368421052631576
Precision:  0.5742690058479532
Recall:  0.47368421052631576
F-1 measure:  0.5050125313283209

Confusion Matrix:

[[ 4  6]
 [14 14]]
              precision    recall  f1-score   support

          -1       0.22      0.40      0.29        10
           1       0.70      0.50      0.58        28

    accuracy                           0.47        38
   macro avg       0.46      0.45      0.43        38
weighted avg       0.57      0.47      0.51        38



## Report


In the workshop we have successfully applied the web scraping sentimental analysis methods. In the web scraping process, we have collected the reviews of 6 movies from the IMDB site. To scrap the data from a website we have used BeautifulSoup and requests library. To do the sentimental analysis we have used NLTK library. Using the Tf-idf scheme, the accuracy of SVM model is 0.74 and for Decision tree model is 0.5. while through the BoW scheme, the accuracy of SVM model is 0.87 and for Decision tree model is 0.47. The accuracy of Decision tree model is very low and the reason is data insufficient. The BoW Scheme is performing well in terms of accuracy. 